pkg/listwatch/listwatch.go

```diff
@@ -135,24 +137,27 @@ type multiListerWatcher []cache.ListerWatcher
 // a single result.
 func (mlw multiListerWatcher) List(options metav1.ListOptions) (runtime.Object, error) {
 	l := metav1.List{}
-	var resourceVersions []string
-	for _, lw := range mlw {
+	resourceVersions := make([]string, len(mlw))
+	for i, lw := range mlw {
 		list, err := lw.List(options)
 		if err != nil {
-			return nil, err
+			klog.Errorf("Failed to list resources: %v", err)
+			continue
 		}
 		items, err := meta.ExtractList(list)
 		if err != nil {
-			return nil, err
+			klog.Errorf("Failed to extract list: %v", err)
+			continue
 		}
 		metaObj, err := meta.ListAccessor(list)
 		if err != nil {
-			return nil, err
+			klog.Errorf("Failed to list accessor: %v", err)
+			continue
 		}
 		for _, item := range items {
 			l.Items = append(l.Items, runtime.RawExtension{Object: item.DeepCopyObject()})
 		}
-		resourceVersions = append(resourceVersions, metaObj.GetResourceVersion())
+		resourceVersions[i] = metaObj.GetResourceVersion()
```

Comment:
Since we continue in case of errors, we need to switch from appending to to using an indexed based accessor. Otherwise, the resource versions will get out of sync with the listers used to get the resource.

Reply:
I wonder if this change has any impact on the watchers. Originally, the resource versions provided to the watch functions were always valid, but now when the `List` operation fails an empty resource version is provided to the `Watch` function.

As far as I can tell, the bug originally reported will still exist with the watchers since whenever a watch operation fails, the whole multi lister watcher fails [1]. Now that the watch functions are passed incorrect resource versions (`""` in case the `List` failed), failures in the `Watch` operation should occur more often than what it used to, so we will end up with the same issue but somewhere else. 

[1] https://github.com/kubernetes/kube-state-metrics/blob/master/pkg/listwatch/listwatch.go#L210-L214

Reply:
This is a good catch! Do you see any problems with modifying the multiWatch to not fail when an individual watcher fails?

Reply:
I don't see any problem with that. However, I am wondering if the failed Listers should really be propagated to the watchers. Theoretically, if the `list` operation failed, I wouldn't expect us to try to setup a watcher for the failed listwatcher, but with the way, the code is currently, that's what we would be doing.

Reply:
I am not sure how we would do that. The `List` and `Watch` functions operate on a `multiListerWatcher` which is an alias for `[]cache.ListerWatcher`. There are different reasons why a `List` operation can fail other than permissions issues, for example there could be a temporary 5xx error from the kube-apiserver. If the `List` function modifies the slice it operates on, a relist operation will never take into account the failed list watchers.

Reply:
Another thing I noticed was that the ResourceVersion which the `List` function modifies actually persists between relists. When watching two namespaces, this leads to errors in the form of: 

```
Failed to watch *v1.Pod: failed to list *v1.Pod: resourceVersion: Invalid value: "1829687/1829687": strconv.ParseUint: parsing "1829687/1829687": invalid syntax
```
This is because the resource version for the list option is set [here](https://github.com/kubernetes/kube-state-metrics/blob/master/pkg/listwatch/listwatch.go#L159) and is never reset.

I believe this might be causing https://github.com/kubernetes/kube-state-metrics/issues/694.

Reply:
Will resolve this since the file was removed.